src/App/components/Section/components/LaneBlock/LaneBlock.tsx

```diff
@@ -0,0 +1,163 @@
+import React, { useEffect, useMemo, useRef, useState } from 'react';
+import clsx from 'clsx';
+import { nanoid } from 'nanoid';
+import { useSelector } from 'react-redux';
+
+import generateTime from '../../../../utils/generateTime';
+import useWindowDimensions from '../../../../hooks/useWindowDimensions';
+import useTimeInfo from '../../../../hooks/useTimeInfo';
+import useTheme from '../../../../hooks/useTheme';
+import { ILaneBlockProps } from './LaneBlock.types';
+import { IInitialState } from '../../../../redux/types';
+
+import style from './LaneBlock.module.scss';
+
+const LaneBlock: React.FC<ILaneBlockProps> = ({ location }) => {
+  const {
+    locations: { userLocation }
+  } = useSelector((state: IInitialState) => state);
+  const [activeIndex, setActiveIndex] = useState(0);
+  const timeInfo = useTimeInfo(location);
+  const bodyTheme = useTheme(style.lightBody, style.darkBody);
+  const nonWorkingHours = useTheme(style.nonWorkHoursLight, style.nonWorkHoursDark);
+  const { width } = useWindowDimensions();
+  const isMobileView = width <= 600;
+
+  const isUserLocation = useMemo(() => {
+    return location?.city === userLocation?.city && location?.lat === userLocation?.lat;
+  }, [userLocation?.city, userLocation?.lat, location?.city, location?.lat]);
+
+  const userTimeInfo = useTimeInfo(userLocation);
+  const activeHour =
+    Number(userTimeInfo.minutes) > 30 ? Number(userTimeInfo.hours) + 1 : Number(userTimeInfo.hours);
+  const activeMinutes = Number(userTimeInfo.minutes) > 30 ? '00' : '30';
+  const activeTime = `${activeHour < 10 ? '0' + activeHour : activeHour}:${activeMinutes}`;
+
+  const timeline = generateTime(24, 30, 0, 23);
+  const newTimeline = timeline.map(value => {
+    const times = value.split(':');
+    const offsetTime = String((timeInfo.offsetTime as number) / 60).split('.');
+    let offsetHour = Number(offsetTime[0]);
+    let offsetMinute = Number(offsetTime[1]) * (600 / 100);
+    if (isNaN(offsetMinute)) {
+      offsetMinute = 0;
+    }
+    if (offsetMinute === 450) {
+      offsetHour = offsetHour + 1;
+      offsetMinute = 0;
+    }
+    let hours = Number(times[0]) + offsetHour;
+    let minutes = Number(times[1]) + offsetMinute;
+    if (minutes === 60) {
+      hours = hours + 1;
+      minutes = 0;
+    }
+    if (hours < 0) {
+      hours = hours + 24;
+    }
+    if (hours > 23) {
+      hours = hours - 24;
+    }
+    return `${hours < 10 ? '0' + hours : hours}:${minutes < 30 ? '00' : '30'}`;
```

Comment:
It would be better to separate this piece of logic into another function maybe create a utils folder and generateTimeTable.tsx and there write all the logic. It will increase of readability of the component.

Also, I think we should memoize the logic to avoid re-renders.

Reply:
I have done this now